pkg/apis/machine/v1beta1/machinehealthcheck_types.go

```diff
@@ -80,6 +80,15 @@ type MachineHealthCheckSpec struct {
 	// +kubebuilder:validation:Pattern="^([0-9]+(\\.[0-9]+)?(ns|us|µs|ms|s|m|h))+$"
 	// +kubebuilder:validation:Type:=string
 	NodeStartupTimeout metav1.Duration `json:"nodeStartupTimeout,omitempty"`
+
+	// RemediationTemplate is a reference to a remediation template
```

Comment:
How will the controller be able to create new objects from the template? Who is going to be adding the appropriate RBAC? Is that going to be part of the default set of RBAC or will another component grant permission to the MHC service account in their own way?

Reply:
There was no change in this logic from upstream.
In a nutshell the existence of a template means that the proper CRD exist so a CR (EMR) can be created.
If you want to read more, [here](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20191030-machine-health-checking.md) is the upstream md file

Reply:
Just because a CRD exists, doesn't mean a user has permission to create resources 😉  Something will need to give the service account for the Machine Health Check controller permissions to be able to create these resources, it could be that we just add those to the defaults (which I think is perfectly acceptable), but to do this we need to know ahead of time all of the types that could possible be referred to here

Reply:
I understand.
I actually had to do it on my e2e test.
However I'm not sure if it is something we can know ahead - since I'm pretty sure that the CRDs are defined by the third  party (external).
It might be something that needs to be updated after coordination with the external user.
/cc @n1r1  what do you think ?

Reply:
Good point.
Unfortunately we can not list all the possible resources as some of them may still not exist and we want this to be generic and plugable (such that new remediation provider could add its own logic without any change in MHC/MAO)

I believe the way it works upstream is that the remediation template and remediation request are defined with apiGroup that already exists in their RBAC, and with `resources: '*'`.

I think we might want to go in a slightly different way. We could create a specific apiGroup that is only intended for external remediation, e.g. `ext-remediation.machine.openshift.io` and update the RBAC to allow creation of any resource with that apiGroup.

WDYT?

Reply:
I guess this comes down to, how many do we realistically expect to support within OpenShift? If it's 2/3, then we can probably add those to the RBAC in this repo as we invent them. If it's more, perhaps the operators responsible for those external remediation strategies should be responsible for adding a binding for the service account?

Eg, could the cluster-baremetal-operator add a role/binding to give the correct permissions if the platform is baremetal and it's deploying some external remediation controller

Also, are external remediation templates always expected to be in a particular namespace? Or would they vary across a cluster/per external remediation implementation? This could affect how we want to handle the RBAC

Reply:
> I guess this comes down to, how many do we realistically expect to support within OpenShift? If it's 2/3, then we can probably add those to the RBAC in this repo as we invent them. If it's more, perhaps the operators responsible for those external remediation strategies should be responsible for adding a binding for the service account?

Yeah, but this creates unnecessary coupling. Once we get this PR merged, a new external remediation controllers could be installed via OLM and just work, without any need to update something in core OCP. 
What would be the downside to implement the suggestion I made?

> Eg, could the cluster-baremetal-operator add a role/binding to give the correct permissions if the platform is baremetal and it's deploying some external remediation controller

From my point of view, external remediation is a generic infrastructure. It is not tied to a specific platform. 
I prefer to have a complete decoupling between MHC and the external remediation controller.
There's a known api, and whoever wants can just implement it.

> Also, are external remediation templates always expected to be in a particular namespace? Or would they vary across a cluster/per external remediation implementation? This could affect how we want to handle the RBAC

Yes. They are required to be in the namespace of the unhealthy machine (which is `openshift-machine-api` in this case)

Reply:
> What would be the downside to implement the suggestion I made?

IMO forcing all external remediations into a single API group will limit us severely. This means that only operators built by OpenShift and maintained by OpenShift would be able to be used, ie, no community remediation templates, if that's not an issue then sure, we can go down this route

> If it's more, perhaps the operators responsible for those external remediation strategies should be responsible for adding a binding for the service account?

I think my preference would be this kind of approach, it would be consistent across both openshift and community remediation methods at least

> From my point of view, external remediation is a generic infrastructure. It is not tied to a specific platform.
I prefer to have a complete decoupling between MHC and the external remediation controller.
There's a known api, and whoever wants can just implement it.

Yes I appreciate that and definitely want it to be generic, but that doesn't fit RBAC very well, so we need to come up with some solution of allowing the MHC permission for the groups that are appropriate.

Reply:
> > What would be the downside to implement the suggestion I made?
> 
> IMO forcing all external remediations into a single API group will limit us severely. This means that only operators built by OpenShift and maintained by OpenShift would be able to be used, ie, no community remediation templates, if that's not an issue then sure, we can go down this route

Good point.
We could take an hybrid approach, where openshift remediators can be seamlessly added while community operators would need to do some extra step (either patch the SA or update the RBAC in this repo)

> 
> > If it's more, perhaps the operators responsible for those external remediation strategies should be responsible for adding a binding for the service account?
> 
> I think my preference would be this kind of approach, it would be consistent across both openshift and community remediation methods at least

The service account could be different in different distributions, isn't it? for example, MHC SA could be `ocp-mhc-sa` and in CAPI based clusters it could be `capi-mhc-sa`, so the remediator will need to patch all "known" SAs.

Also, I'm not sure if someone will be able to remember that changing MHC service account breaks other external components.

Maybe remediators should expose some config value. it could be either the `apiGroup` to be used or the `ServiceAccount` to be patched.

Reply:
a new idea came up to my mind - what about just having a non openshift api group which is specifically for external remediation?
for example `external-remediation.io`. We could propose that upstream as well, such that community remediators could work the same on OCP and upstream, by using that group.

WDYT?

Reply:
just wanted to add another idea here. we could create the default build with a tightly focused api group for the external remediation template, but give the user clear instructions on how to override/expand that initial group. basically as a day 2 operation they could create their own special EMRs and load those in, but by default it would work "out of the box" with a reasonable setting.

> We could take an hybrid approach, where openshift remediators can be seamlessly added while community operators would need to do some extra step (either patch the SA or update the RBAC in this repo)

+1, i think this is saying something similar

this situation is also similar to how we have modified the cluster-api provider in the cluster-autoscaler. it's a different problem there, but perhaps a similar solution. we compile the autoscaler to know about our resrouce groups, but it is possible for the autoscaler to be reconfigured by environment variable to use a different api group. this allows us to share code with upstream capi and also keep our openshift types for when we deploy.

Reply:
Thanks @elmiko 

We discussed this issue with the CAPI community, and CAPI is already using [RBAC aggregated label](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#aggregated-clusterroles) which allows external components to add their RBAC rules.

I believe this is an elegant way to solve this problem and it's always good to be aligned with upstream.